test/gpu_test.cpp

```diff
@@ -160,3 +164,105 @@ TEST(gpu_test, memory_manager_seperate_workspaces_allocsize)
     emitter.allocate_primitive_memory();
     EXPECT_EQ(emitter.sizeof_device_allocation(), total_size);
 }
+
+TEST(gpu_test, topk_3d_global_mem)
+{
+    Shape shape{4, 3073, 5};
+    auto A_cpu = make_shared<op::Parameter>(element::f32, shape);
+    auto A_gpu = make_shared<op::Parameter>(element::f32, shape);
+
+    auto B_cpu = make_shared<op::TopK>(A_cpu, 1, element::i64, 10, true);
+    auto B_gpu = make_shared<op::TopK>(A_gpu, 1, element::i64, 10, true);
+
+    auto cpu_f_0 = make_shared<Function>(make_shared<op::GetOutputElement>(B_cpu, 0),
+                                         op::ParameterVector{A_cpu});
+    auto cpu_f_1 = make_shared<Function>(make_shared<op::GetOutputElement>(B_cpu, 1),
+                                         op::ParameterVector{A_cpu});
+
+    auto gpu_f_0 = make_shared<Function>(make_shared<op::GetOutputElement>(B_gpu, 0),
+                                         op::ParameterVector{A_gpu});
+    auto gpu_f_1 = make_shared<Function>(make_shared<op::GetOutputElement>(B_gpu, 1),
+                                         op::ParameterVector{A_gpu});
+
+    test::Uniform<float> rng(0.0f, 1.0f);
+    vector<vector<float>> args_index;
+
+    for (shared_ptr<op::Parameter> param : cpu_f_0->get_parameters())
+    {
+        vector<float> tensor_val(shape_size(param->get_shape()));
+        rng.initialize(tensor_val);
+        args_index.push_back(tensor_val);
+    }
+
+    auto cpu_results_0 = execute<float, int64_t>(cpu_f_0, args_index, "CPU");
+    auto gpu_results_0 = execute<float, int64_t>(gpu_f_0, args_index, "GPU");
+
+    for (size_t i = 0; i < gpu_results_0.size(); i++)
+    {
+        EXPECT_TRUE(test::all_close(gpu_results_0.at(i), cpu_results_0.at(i)));
```

Comment:
This is one of the few cases where `all_close` is wrong. When you rearrange a list of floats, they won't change value. If they do, something is very wrong.
I just checked -- ONNX topk, which this is implementing, specifies you order by (-value, index), so all backends should agree. Here you should be checking against INTERPRETER (reference implementation) rather than CPU. Then we can move the test into the backend-independent tests. We might need to fix the reference implementation first.
https://github.com/onnx/onnx/blob/master/docs/Operators.md#TopK

Reply:
For reference, neither TensorFlow nor MXNet specify what to do on a tie.